### PR TITLE
gRPC service

### PR DESCRIPTION
**Pull request by jsanda**
This PR is for #137 and is in place of #141.

### CODE REVIEW

File: VERSION
Comment (adejanovski) [inline VERSION:1]: You'll need a little rebase before merging.

File: medusa/config.py
Comment (adejanovski) [inline medusa/config.py:133]: Could the value be moved to a constant?

File: medusa/restore_node.py
Comment (adejanovski) [inline medusa/restore_node.py:88]: This means the gRPC server can exclusively be used for k8s clusters? I thought we would have the option to use it for classic clusters as well and ease automation for folks who prefer invoking Medusa remotely.
Maybe we should have another config setting to indicate whether or not we're dealing with k8s.

File: medusa/restore_node.py
Comment (adejanovski) [inline medusa/restore_node.py:92]: Same here, this should be subject to running into k8s or not instead of grpc or not.

File: medusa/storage/google_storage.py
Comment (adejanovski) [inline medusa/storage/google_storage.py:33]: This doesn't belong in this PR. Either we have a missing commit in `master` or this PR should be properly rebased against master.
I'll check what's wrong here (and the rebase is needed anyway).

File: setup.py
Comment (adejanovski) [inline setup.py:62]: Since the code is there and it is imported, won't the grpc requirements be needed in all scenarios?

**Review by adejanovski — APPROVED** [review]
I have a few comments around the fact that the grpc server seems to be exclusively available for k8s while users could benefit from it even on non dockerized environments.
I'll let @arodrime run some tests but I'm ok with merging the PR if it gets properly rebased against master and commits get squashed into a single one.

File: medusa/storage/google_storage.py
Comment (arodrime) [reply medusa/storage/google_storage.py:33]: @jsanda - Alex understood where this was coming from and fixed it. Rebase should do now that we updated master.

File: medusa/restore_node.py
Comment (jsanda) [reply medusa/restore_node.py:88]: > I thought we would have the option to use it for classic clusters as well and ease automation for folks who prefer invoking Medusa remotely.

Absolutely. There is nothing Kubernetes specific about the gRPC service. There are integration tests that exercise it, and they do not run in Kubernetes.

> Maybe we should have another config setting to indicate whether or not we're dealing with k8s.

This would be my preference as well. I will add another config setting.

Comment (emerkle826) [reply medusa/restore_node.py:88]: I've added a K8s configuration setting now and changed the logic to trigger based on that.

File: medusa/config.py
Comment (emerkle826) [reply medusa/config.py]: done

File: medusa/restore_node.py
Comment (emerkle826) [reply medusa/restore_node.py]: done

File: setup.py
Comment (emerkle826) [reply setup.py]: Indeed. Moved them to install_requires

**Comment by emerkle826** [comment]
@adejanovski I would appreciate a re-review of this. I believe I have addressed all the comments so far.

File: medusa/service/__init__.py
Comment (adejanovski) [inline medusa/service/__init__.py:2]: Wrong copyright

File: medusa/service/grpc/__init__.py
Comment (adejanovski) [inline medusa/service/grpc/__init__.py:2]: Wrong copyright

File: medusa/service/grpc/client.py
Comment (adejanovski) [inline medusa/service/grpc/client.py:1]: Missing copyright/license

File: medusa/service/grpc/restore.py
Comment (adejanovski) [inline medusa/service/grpc/restore.py:1]: Missing copyright/license

File: medusa/service/grpc/server.py
Comment (adejanovski) [inline medusa/service/grpc/server.py:1]: Missing copyright/license

File: medusa/utils.py
Comment (adejanovski) [inline medusa/utils.py:2]: Wrong copyright

**Review by adejanovski — CHANGES_REQUESTED** [review]
@jsanda @emerkle826,

we're very close to merging this PR. There's still a code path which is specific to grpc while it should be specific to k8s instead on taking snapshots with Jolokia.
Also there are some wrong copyrights (should be DataStax Inc. instead of Spotify) and some missing license headers on new files.

**Comment by adejanovski** [comment]
@emerkle826, could you squash your commits into a single one once the requested changes are done?
Thanks!

**Comment by emerkle826** [comment]
> @emerkle826, could you squash your commits into a single one once the requested changes are done?
> Thanks!

Will do.

File: medusa/cassandra_utils.py
Comment (jsanda) [reply medusa/cassandra_utils.py:424]: That is true. We could run the gRPC service along side Cassandra where Medusa has the ability to execute `nodetool`. I agree that this should be a k8s check here.

Comment (emerkle826) [reply medusa/cassandra_utils.py:424]: So, should I just move the jolokia URL to the Kubernetes config, and leave the gRPC config as simply enabled or not? If the issue here is to decouple Jolokia from gRPC, then it seems the Jolokia configuration belongs with the Kubernetes config, not gRPC config.

Comment (adejanovski) [reply medusa/cassandra_utils.py]: Yes, reliance on Jolokia is a requirement for k8s deployments only as Medusa runs in a different container than Cassandra. But we can use gRPC in a non k8s environment where we can run nodetool commands to take snapshots.

**Comment by emerkle826** [comment]
@adejanovski Thanks for the reviews again. I've squashed everything down now. The 2 failing integration tests I believe are due to a bug in the latest Cassandra trunk with SSLContext initialization. I created a ticket here:

https://issues.apache.org/jira/browse/CASSANDRA-16362

Not sure when that will get addressed.

**Comment by jsanda** [comment]
I assume those failures are not specific to changes in this PR? If that is correct, will we be able to merge with those failures?

**Comment by adejanovski** [comment]
@jsanda @emerkle826,

I've pushed a fix for the backup-cluster operation which is not covered (yet) by integration tests and was broken due to a change from this PR.
The failing tests are unrelated to this PR and will not block the merge.

**Review by adejanovski — APPROVED** [review]

**Pull request merged**
